### PR TITLE
107 aead implementation

### PR DESCRIPTION
**Pull request by marsella**
This addresses the first half of #107. It
- Implements encryption and decryption using RustCrypto's ChaCha20Poly1305 library
- Implements all our relevant types as wrappers around the appropriate RustCrypto types
- Implements some basic tests 
- Updates docs to refer to the chosen encryption algorithm

For implementing encryption and decryption, I followed [the basic usage instructions](https://docs.rs/chacha20poly1305/latest/chacha20poly1305/index.html#usage) in the crate. I rejected some alternative implementation options, including:
- disable allocation (for environments without a heap -- doesn't seem relevant for us yet, but maybe we can consider when we start using SGX)
- use a larger nonce (this is a non-standardized algorithm)
- use fewer rounds a la [too much crypto](https://eprint.iacr.org/2019/1492.pdf)

Looking for
- general Rust feedback
- check of cryptography use
- validation of testing; are there more things I should test for?
- organization feedback: should this block of "general" encryption functionality go into a module? The point would be to keep it visually separate from the rest of the crypto module, so that developers are less tempted to... try to use it? And to keep the abstraction layers a little bit more obvious.

### CODE REVIEW

File: dams/src/crypto.rs [hunk 1]
```diff
@@ -36,44 +42,110 @@ impl Default for AssociatedData {
...
+    fn from(associated_data: AssociatedData) -> Self {
+        (&associated_data).into()
+    }
+}
+
+impl From<&AssociatedData> for Vec<u8> {
+    fn from(associated_data: &AssociatedData) -> Self {
+        associated_data.0.as_bytes().to_vec()
+    }
+}
```
[the remark this answers is not in the document]
Comment (picklenerd) [reply dams/src/crypto.rs:62]: I think it would be good to flip the implementation. You should be able to get a `Vec<u8>` from a `String` directly with [String::into_bytes](https://doc.rust-lang.org/std/string/struct.String.html#method.into_bytes). That would let the non-reference version get the data without copying it and then the reference version could just clone the data and call the non-reference version.

I think having both is fine for convenience if you're using the reference version a lot. I would prefer having the caller do `.clone().into()` just so the cloning is explicit, but that adds noise so doing it this way will probably look nicer.

File: dams/src/crypto.rs [hunk 2]
```diff
@@ -36,44 +42,110 @@ impl Default for AssociatedData {
...
 
 #[allow(unused)]
 impl<T> Encrypted<T>
 where
-    T: From<Bytes>,
-    Bytes: From<T>,
+    T: From<Vec<u8>>,
+    Vec<u8>: From<T>,
```
[the remark this answers is not in the document]
Comment (picklenerd) [reply dams/src/crypto.rs:94]: Totally agree. This will be much easier to work with.

File: dams/src/crypto.rs [hunk 3]
```diff
@@ -251,30 +348,60 @@ mod test {
...
+    }
+
     #[test]
-    #[should_panic(expected = "not yet implemented")]
-    fn encryption_not_implemented() {
-        let bytes = Bytes::default();
-        let _encrypted = Encrypted::encrypt(&EncryptionKey, bytes, &AssociatedData::default());
+    fn encryption_not_obviously_broken() {
+        let (bytes, encrypted_bytes, _) = encrypt_random_bytes();
+        assert_ne!(bytes, encrypted_bytes.ciphertext);
```
Comment (indomitableSwan) [inline dams/src/crypto.rs:371]: I would also check that the ciphertext object contains the expected associated data.

File: dams/src/crypto/client.rs [hunk 4]
```diff
@@ -16,7 +16,8 @@ use super::{Encrypted, Secret};
 /// It takes the following steps:
 /// 1. Derive a master key from the [`OpaqueExportKey`]
 /// 2. Generate a new [`StorageKey`] to encrypt stored data with
-/// 3. Encrypt the storage key with the master key
+/// 3. Encrypt the storage key under the master key, using the ChaCha20Poly1305
```
Comment (indomitableSwan) [inline dams/src/crypto/client.rs:19]: Same comment  as elsewhere re: specifying chacha20poly1305.

File: dams/src/crypto/client.rs [hunk 5]
```diff
@@ -31,7 +32,8 @@ pub fn create_and_encrypt_storage_key(
 ///
 /// This must be run by the client. It takes the following steps:
 /// 1. Generates a new secret
-/// 2. Encrypt it under the [`StorageKey`]
+/// 2. Encrypt it under the [`StorageKey`], using the ChaCha20Poly1305 AEAD
```
Comment (indomitableSwan) [inline dams/src/crypto/client.rs:35]: Same comment as above.

**Review by indomitableSwan — CHANGES_REQUESTED** [review]
Thanks, Marcella!

- Most of the comments are nits.
- There are a couple inline comments on testing and also my high-level comments on this topic below.
- There are a couple of questions on how we are using the chacha20poly1305 dependency. I may be misunderstanding something, but it looks like we might be re-implementing some functionality that the dependency should already provide.

Responses to (two of) your high-level questions:
> validation of testing; are there more things I should test for?

Yes, I think testing needs to be expanded. We had some more nuanced and complete unit testing for zkChannels, iirc, which might provide a good reference point to start. But:
- organizing the tests so that we can try them multiple times on some randomly generated input, e.g., combine encryption & decryption working tests into one and run on a test vector
- checking more carefully that associated data is working as intended, e.g., making sure the AD in the encrypted object is exactly what is expected, making sure you can add to the default AD without anything weird happening, making sure that minor tweaks to the AD causes the test to fail, etc.
- consider tests with non-randomly generated messages.
- consider testing that tweaking the ciphertext by flipping a bit somewhere in the encrypted part of the payload causes decryption to fail.

> organization feedback: should this block of "general" encryption functionality go into a module? The point would be to keep it visually separate from the rest of the crypto module, so that developers are less tempted to... try to use it? And to keep the abstraction layers a little bit more obvious.

I think the fact that you are asking this question means we probably haven't hit on the right level of visibility for the crypto module within the rest of the project. But I don't think we should try to fix this here.

File: dams/src/crypto.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (marsella) [reply dams/src/crypto.rs:62]: Thanks, that makes sense! I only call `from` on a reference once, so I'll put the clone in and remove the extra `From` impl.

File: dams/src/crypto.rs [hunk 6]
```diff
@@ -36,44 +42,110 @@ impl Default for AssociatedData {
...
+/// A ciphertext representing an object of type `T`, encrypted under the
+/// ChaCha20Poly1305 AEAD scheme.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct Encrypted<T> {
-    ciphertext: Bytes,
+    ciphertext: Vec<u8>,
     associated_data: AssociatedData,
+    nonce: chacha20poly1305::Nonce,
     original_type: PhantomData<T>,
 }
```
[the remark this answers is not in the document]
Comment (marsella) [reply dams/src/crypto.rs:72]: I like the way that the current name reads -- e.g. it's an encrypted secret -- but I'm flexible. I think @vidyaap is actively using this type right now, so maybe she has an opinion?

We discussed `PhantomData<T>` offline. To recap, it's a way of tying an encrypted blob to the type that was originally encrypted. The `PhantomData` type lets us make it a compile-time guarantee that an encrypted `T` will always decrypt to a `T`, and those decrypted bytes won't be coerced into some other type.

File: dams/src/crypto.rs [hunk 7]
```diff
@@ -36,44 +42,110 @@ impl Default for AssociatedData {
...
 
-/// A well-formed symmetric encryption key for the  authenticated encryption
-/// with associated data (AEAD) scheme (TODO #107: specify encryption scheme).
+/// A well-formed symmetric encryption key.
+///
+/// This key is for [ChaCha20Poly1305](https://www.rfc-editor.org/rfc/rfc8439),
+/// an authenticated encryption with associated data (AEAD) scheme, using the
+/// recommended 20 rounds.
```
[the remark this answers is not in the document]
Comment (marsella) [reply dams/src/crypto.rs:78]: Thanks, this is a good call. I will consolidate all the scheme-specific stuff under `Encrypted` as you suggest.

File: dams/src/crypto.rs [hunk 8]
```diff
@@ -36,44 +42,110 @@ impl Default for AssociatedData {
...
+        // Set up cipher with key
+        let cipher = ChaCha20Poly1305::new(&enc_key.0);
+
+        // Format plaintext and associated data
+        let nonce = ChaCha20Poly1305::generate_nonce(rng);
+        let ad_vec: Vec<u8> = associated_data.into();
+        let payload = Payload {
+            msg: &Vec::from(object),
+            aad: &ad_vec,
+        };
```
[the remark this answers is not in the document]
Comment (marsella) [reply dams/src/crypto.rs:114]: Discussed offline, but to recap: Nope, the AEAD trait will automatically coerce bytes (`Vec<u8>`) to a payload with no associated data, but [the tests suggest](https://github.com/RustCrypto/AEADs/blob/master/chacha20poly1305/tests/lib.rs#L26-L29) that if you want to use AD, you need to manually create the payload.

If we had some fancier input types (e.g. a struct that included fields that would go into the plaintext and other fields for the AD), we could implement `Into<Payload<'msg, 'aad>` for it to make this bit cleaner. But since we only do the conversion once and we don't have a all-inclusive input type, I think this is the best option.

File: dams/src/crypto.rs [hunk 9]
```diff
@@ -139,15 +211,15 @@ struct MasterKey;
 
 #[allow(unused)]
 impl MasterKey {
-    /// Encrypt the given [`StorageKey`] under an AEAD scheme (TODO #107:
-    /// describe).
+    /// Encrypt the given [`StorageKey`] under the [`MasterKey`], using the
+    /// ChaCha20Poly1305 AEAD scheme.
```
[the remark this answers is not in the document]
Comment (marsella) [reply dams/src/crypto.rs:215]: I will remove it here, this is a private type so anyone who can see this doc will also be able to see that a master key is just a wrapper for an `EncryptionKey` and can click through to see the scheme.

File: dams/src/crypto.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vidyaap) [reply dams/src/crypto.rs]: I think `Encrypted` makes sense if `ciphertext` is only one field in it (as opposed to calling it `Ciphertext` and it still having a `ciphertext` field).

**Comment by marsella** [comment]
I've addressed all your comments and added some of the new tests you requested:
- I added iteration to the randomness-based tests, including basic encryption/decryption, decryption failing with the wrong key, and tweaking the ciphertext
- I am not going to add associated data tests yet because I haven't written the AD interface. I will add those in the next PR.
- I couldn't think of any particular edge cases that might cause enc/dec to fail, so I didn't add those. I could switch to a seeded RNG to make the tests reproducible; we had a long discussion about this last time and I thought it basically came down to preference whether to use seeded RNGs or not.
- I added the "tweak a ciphertext" test

I also added some tests on nonces: a check that they are not obviously repeated, and a check that changing the nonce fails decryption.

**Review by indomitableSwan — APPROVED** [review]
LGTM.
